wa/instruments/perf.py

```diff
@@ -13,126 +13,220 @@
+                        inputs (``event``, ``pid``, ...);
+                      - ``args`` (``str`` or valid command): the post-``--``
+                        arguments. This is typically the command ``perf`` will
+                        launch and monitor. Therefore, a valid command
+                        dictionary (same as this one) is accepted;
+
+                  As an example, the default behaviour can be replicated
+                  through::
+
+                      :language: yaml
+
+                      perf:
+                          commands:
+                              default_behaviour:
+                                  command: stat
+                                  flags:
+                                      - all-cpus
+                                  kwflags:
+                                      event:
+                                          - migrations
+                                          - cs
+                                  args:
+                                      command: sleep
+                                      args: 1000
+                                  stderr: '&1'
+                                  stdout: stat.out
+                 """),
+        Parameter('post_commands', kind=YamlCommandDescriptor, default=None,
+                  description="""
+                  Dictionary of commands to be run after the workloads run
+                  (same format as ``commands``).
+                 """),
     ]
 
     def __init__(self, target, **kwargs):
         super(PerfInstrument, self).__init__(target, **kwargs)
         self.collector = None
 
     def initialize(self, context):
+        # pylint: disable=unused-argument
+        # pylint: disable=access-member-before-definition
+        # pylint: disable=attribute-defined-outside-init
+        if self.commands is None:
+            if isinstance(self.optionstring, str):
+                self.optionstring = [self.optionstring]
+
+            if not self.labels:
+                self.labels = [f'default{i}'
+                               for i, _ in enumerate(self.optionstring)]
+            elif isinstance(self.labels, str):
+                self.labels = [self.labels]
+
+            if len(self.labels) != len(self.optionstring):
+                raise ValueError("Lengths of labels and optionstring differ")
+
+            self.commands = YamlCommandDescriptor({
+                label: {
+                    'command': 'stat',
+                    'kwflags': {'event': self.events},
+                    'options': options,
+                    'args': {
+                        'command': 'sleep',
+                        'args': 1000,
+                    },
+                    'stderr': '&1',
+                    'stdout': 'stat.out',
+                }
+                for label, options in zip(self.labels, self.optionstring)
+            })
+        # TODO: else, detect if events or optionstring were passed and error?
```

Comment:
Is there a standard way for checking that a `Parameter` was passed through the agenda or not? I would like to `raise` and error if the user attempts to use any of `labels`, `optionstring` or `events` when they've passed `command`. I was thinking of testing these variables against their defaults but that doesn't detect the case where an agenda uses incorrectly these parameters while passing their default values ...

Reply:
We don't have a mechanism for this at the moment. We typically deal with this by not specifying a `default` inside the prameter, but setting it inside `__init__` if the value is `None`.

Now that I think about it, adding such a mechanism should be straight forward.

Reply:
Erm, actually, turns out it's not that straight forward. Never mind.

Reply:
Done (a bit hacky, though)